### ISSUE TITLE
--tests seems to initialize every test causing slow performance

### ISSUE THREAD

**Issue by MattNohelty** [gradle/gradle#3851]
The "--tests" command line option, used to run a subset of tests, seems to initialize every test, not just the ones specified in the filter parameter.  It only runs the tests that are specified so it's technically working correctly but the fact that it initializes every test causes a significant performance hit in a large application with many tests.  

This behavior does not seem to happen when using "-Dtest.single".  In this case, only the specified tests are initialized and run so this runs significantly faster than "--tests".  In my large enterprise application at work, running a set of tests from one class with "-Dtest.single" completes in less than 30 seconds but the exact same tests take over 70 seconds when run with "--tests".  Based on debug logging, it appears this extra time is all spent initializing tests that never end up being run.

I would expect "--tests" to only initialize the tests that are specified in the filter parameter, not every test.  In other words, this should do exactly what "-Dtest.single" does.
 
We could just use "-Dtest.single", even though it sounds like it's being deprecated, but most of our developers use Intellij and run their tests inside the IDE and delegate the test execution to Gradle.  Intellij uses "--tests" by default and I don't see any way to change behavior. This is creating major frustration for our developers because running a single test takes significantly longer if we delegate the builds/tests to Gradle compared to running them with Intellij's test runner and the time difference is seems to largely be due to this issue.

I created a quick sample project with two classes, Foo.java and Bar.java. Each class has one method and a corresponding test class with a single test.  Below is a snippet from the debug logs if I run the tests for Foo in both ways described above.

**gradlew cleanTest test --tests "com.test.gradletest.TestFoo" --debug**

> Line 865: 23:15:50.211 [DEBUG] [TestEventLogger] 
Line 866: 23:15:50.211 [DEBUG] [TestEventLogger] Gradle Test Run :test STARTED
Line 887: 23:15:50.818 [DEBUG] [TestEventLogger] 
Line 888: 23:15:50.818 [DEBUG] [TestEventLogger] Gradle Test Executor 1 STARTED
Line 890: 23:15:50.823 [DEBUG] [TestEventLogger] 
Line 891: 23:15:50.823 [DEBUG] [TestEventLogger] com.test.gradletest.TestBar STARTED
Line 892: 23:15:50.894 [DEBUG] [TestEventLogger] 
Line 894: 23:15:50.894 [DEBUG] [TestEventLogger] com.test.gradletest.TestBar PASSED
Line 895: 23:15:50.894 [DEBUG] [TestEventLogger] 
Line 896: 23:15:50.895 [DEBUG] [TestEventLogger] com.test.gradletest.TestFoo STARTED
Line 897: 23:15:50.904 [DEBUG] [TestEventLogger] 
Line 898: 23:15:50.904 [DEBUG] [TestEventLogger] com.test.gradletest.TestFoo > testFoo STARTED
Line 899: 23:15:50.907 [DEBUG] [TestEventLogger] 
Line 900: 23:15:50.907 [DEBUG] [TestEventLogger] com.test.gradletest.TestFoo > testFoo PASSED
Line 901: 23:15:50.907 [DEBUG] [TestEventLogger] 
Line 902: 23:15:50.907 [DEBUG] [TestEventLogger] com.test.gradletest.TestFoo PASSED
Line 903: 23:15:50.908 [DEBUG] [TestEventLogger] 
Line 904: 23:15:50.908 [DEBUG] [TestEventLogger] Gradle Test Executor 1 PASSED
Line 910: 23:15:50.928 [DEBUG] [TestEventLogger] 
Line 911: 23:15:50.929 [DEBUG] [TestEventLogger] Gradle Test Run :test PASSED

**gradlew cleanTest test -Dtest.single=TestFoo --debug**
>23:51:59.000 [DEBUG] [TestEventLogger] 
23:51:59.021 [DEBUG] [TestEventLogger] Gradle Test Executor 1 STARTED
23:51:59.024 [DEBUG] [TestEventLogger] 
23:51:59.024 [DEBUG] [TestEventLogger] com.test.gradletest.TestFoo STARTED
23:51:59.231 [DEBUG] [TestEventLogger] 
23:51:59.231 [DEBUG] [TestEventLogger] com.test.gradletest.TestFoo > testFoo STARTED
23:51:59.251 [DEBUG] [TestEventLogger] 
23:51:59.251 [DEBUG] [TestEventLogger] com.test.gradletest.TestFoo > testFoo PASSED
23:51:59.253 [DEBUG] [TestEventLogger] 
23:51:59.254 [DEBUG] [TestEventLogger] com.test.gradletest.TestFoo PASSED
23:51:59.257 [DEBUG] [TestEventLogger] 
23:51:59.257 [DEBUG] [TestEventLogger] Gradle Test Executor 1 PASSED
23:51:59.337 [DEBUG] [TestEventLogger] 
23:51:59.337 [DEBUG] [TestEventLogger] Gradle Test Run :test PASSED

As you can see, the "--tests" option does some work on TestBar while "-Dtest.single" does not.  In a small example like this, it's not noticeable but in a project with thousands of tests, this extra bit of overhead for every test adds up to a significant performance hit.

**Comment by 4n00p** [comment]
This has been reported before, and is particularly bad when the forkEvery option is used.
[https://discuss.gradle.org/t/test-run-slow-when-running-with-test-com-example-footest/18489/2](https://discuss.gradle.org/t/test-run-slow-when-running-with-test-com-example-footest/18489/2)

**Comment by MattNohelty** [comment]
Thanks for the info @4n00p. Looks like that discussion was from over a year ago so I'm guessing that didn't really go anywhere.  I'm a bit surprised there isn't more noise about this.  We are working on converting our builds over from ant and this is causing a real impact for us in terms of developer productivity. We have about 3000 tests and as far as I can tell we aren't doing anything out of the ordinary and are not using forkEvery. Seems like this would be very noticeable issue for any larger project.

**Comment by blindpirate** [comment]
Well, what `-Dtest.single` does is simple concatenating a pattern string: https://github.com/gradle/gradle/blob/f71d8d69a1f1e031d6ffd105664da0ce9440f04b/subprojects/plugins/src/main/java/org/gradle/api/plugins/JavaBasePlugin.java#L441

```
test.setIncludes(WrapUtil.toSet("**/" + singleTest + "*.class"));
```

However, we cannot simply let `--tests` do that as well, because sometimes we need to filter test methods, and all tests must be loaded to find out the correct matching methods.

We'll discuss this.

**Comment by blindpirate** [comment]
`gradlew cleanTest test --tests "com.test.gradletest.TestFoo"` will no longer load all classes. This will be included in `4.7`.
